### PR TITLE
JProxy update

### PR DESCRIPTION
**Pull request by mkitti**
I have updated the `JProxy` code from @zot from PR #91 . Tests are now passing and there are no conflicts.

### CODE REVIEW

**Comment by zot** [comment]
Thanks a ton for doing this!

File: src/proxy.jl [hunk 1]
```diff
@@ -0,0 +1,1494 @@
...
+
+global useVerbose = false
+global initialized = false
+
+setVerbose() = global useVerbose = true
+clearVerbose() = global useVerbose = false
+
+location(source) = replace(string(source.file), r"^.*/([^/]*)$" => s"\1") * ":" * string(source.line) * ": "
+
+macro verbose(args...)
```
[the remark this answers is not in the document]
Comment (zot) [reply src/proxy.jl:23]: yes, @verbose is just for debugging the project itself and should eventually be removed or commented out. I left some of the logging commented out so I could uncomment it at need (and I did have to uncomment some of them sometimes :) )

File: test/runtests.jl [hunk 2]
```diff
@@ -4,6 +4,7 @@ using JavaCall
 import Dates
 using Base.GC: gc
 
+const pxyptr = JavaCall.pxyptr
```
[the remark this answers is not in the document]
Comment (zot) [reply test/runtests.jl:7]: Correct, I had used that in a println in an earlier version of the test code and I forgot to remove it

File: src/JavaCall.jl [hunk 3]
```diff
@@ -4,7 +4,7 @@ export JavaObject, JavaMetaClass,
        JObject, JClass, JMethod, JString,
        @jimport, jcall, jfield, isnull,
        getname, getclass, listmethods, getreturntype, getparametertypes, classforname,
-       narrow
+       narrow, JProxy, @class, interfacehas, staticproxy
```
[the remark this answers is not in the document]
Comment (zot) [reply src/JavaCall.jl:7]: Maybe "JavaObject", because you use it like a Java object?

File: src/convert.jl [hunk 4]
```diff
@@ -264,14 +280,16 @@ function convert(::Type{@jimport(java.util.List)}, x::Vector, V::Type{JavaObject
...
     pIsCopy = Array{jboolean}(undef, 1)
     buf::Ptr{UInt8} = ccall(jnifunc.GetStringUTFChars, Ptr{UInt8},
-                            (Ptr{JNIEnv}, Ptr{Nothing}, Ptr{jboolean}), penv, jstr.ptr, pIsCopy)
+                            (Ptr{JNIEnv}, Ptr{Nothing}, Ptr{jboolean}), penv, jstr, pIsCopy)
     s = unsafe_string(buf)
     ccall(jnifunc.ReleaseStringUTFChars, Nothing, (Ptr{JNIEnv}, Ptr{Nothing}, Ptr{UInt8}), penv,
-          jstr.ptr, buf)
+          jstr, buf)
     return s
 end
```
Comment (mkitti) [inline src/convert.jl:294]: @aviks Could we incorporate the changes to this file convert.jl into JavaCall proper?

File: src/core.jl [hunk 5]
```diff
@@ -16,8 +17,15 @@ const jdouble = Cdouble
 const jsize = jint
 jprimitive = Union{jboolean, jchar, jshort, jfloat, jdouble, jint, jlong}
 
-struct JavaMetaClass{T}
+mutable struct JavaMetaClass{T}
```
Comment (mkitti) [inline src/core.jl:20]: Why is it necessary to change this to a mutable struct? I haven't look very hard for the answer, but it might save time if someone could explain it.

File: src/jvm.jl [hunk 6]
```diff
@@ -200,6 +201,7 @@ function init(opts)
     jvm = unsafe_load(pjvm)
     global jvmfunc = unsafe_load(jvm.JNIInvokeInterface_)
     global jnifunc = unsafe_load(jnienv.JNINativeInterface_) #The JNI Function table
+    initProxy()
```
Comment (mkitti) [inline src/jvm.jl:204]: We should revert the changes to this file for now.

File: test/Test.java [hunk 7]
```diff
@@ -83,4 +85,47 @@ public String innerString() {
...
+    return "int[]";
+  }
+
+  public String testArrayArgs(int[][] i) {
+    return "int[][]";
+  }
+
+  public String testArrayArgs(Object[] i) {
+    return "java.lang.Object[]";
+  }
```
Comment (mkitti) [inline test/Test.java:130]: The changes to Test.java and the binaries should be reverted for now and moved into a separate sub-package.

File: test/runtests.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkitti) [reply test/runtests.jl:7]: These tests will be moved to a separate sub-package.

File: src/core.jl (hunk 5 above)
Comment (mkitti) [reply src/core.jl:20]: Might it make sense to have mutable and unmutable versions of JavaMetaClass?

Comment (zot) [reply src/core.jl:20]: I'll look around -- I'm pretty sure I had a good reason for that. But maybe not :)

Comment (zot) [reply src/core.jl:20]: Because Java can garbage collect classes and the finalizer on line 84 nulls out ptr for safety. If we can guarantee that deleteref will only ever be called from finalize, it can stay immutable I think.

File: src/convert.jl (hunk 4 above)
Comment (aviks) [reply src/convert.jl:294]: Sure. Is the current code likely a bug?

File: src/JavaCall.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (aviks) [reply src/JavaCall.jl:7]: Yes, that is the most relevant name, the only trouble is that we already have a JavaObject type in JavaCall, so we'll have figure out a transition. However, I think the macro should be called `@java`.

File: src/convert.jl (hunk 4 above)
Comment (mkitti) [reply src/convert.jl:294]: I think it was just quoted out of context. What I meant is that most of the changes to convert.jl seem acceptable to merge. It will have to be updated for the new JNI module though.

File: src/convert.jl [hunk 8]
```diff
@@ -1,19 +1,34 @@
...
+convert(::Type{JavaObject{:int}}, n) = convert(jint, n)
+convert(::Type{JavaObject{:long}}, n) = convert(jlong, n)
+convert(::Type{JavaObject{:byte}}, n) = convert(jbyte, n)
+convert(::Type{JavaObject{:boolean}}, n) = convert(jboolean, n)
+convert(::Type{JavaObject{:char}}, n) = convert(jchar, n)
+convert(::Type{JavaObject{:short}}, n) = convert(jshort, n)
+convert(::Type{JavaObject{:float}}, n) = convert(jfloat, n)
+convert(::Type{JavaObject{:double}}, n) = convert(jdouble, n)
+convert(::Type{JavaObject{:void}}, n) = convert(jvoid, n)
+convert(::Type{JavaObject{T}}, ::Nothing) where T = jnull
```
Comment (mkitti) [inline src/convert.jl:20]: These convert methods seem fine to merge.

File: src/convert.jl [hunk 9]
```diff
@@ -1,19 +1,34 @@
...
-    isnull(obj) && throw(ArgumentError("Cannot convert NULL"))
     realClass = ccall(jnifunc.GetObjectClass, Ptr{Nothing}, (Ptr{JNIEnv}, Ptr{Nothing} ), penv, obj.ptr)
     if isConvertible(T, realClass)  #dynamic cast
-        ptr = ccall(jnifunc.NewLocalRef, Ptr{Nothing}, (Ptr{JNIEnv}, Ptr{Nothing}), penv, obj.ptr)
-        ptr === C_NULL && geterror()
-        return JavaObject{T}(ptr)
+        return JavaObject{T}(obj.ptr)
     end
     throw(JavaCallError("Cannot cast java object from $S to $T"))
 end
```
Comment (mkitti) [inline src/convert.jl:34]: This allows for null objects. We could probably merge this.

File: src/core.jl [hunk 10]
```diff
@@ -82,16 +132,22 @@ isnull(obj::JavaMetaClass) = obj.ptr == C_NULL
 const JClass = JavaObject{Symbol("java.lang.Class")}
 const JObject = JavaObject{Symbol("java.lang.Object")}
 const JMethod = JavaObject{Symbol("java.lang.reflect.Method")}
+const JConstructor = JavaObject{Symbol("java.lang.reflect.Constructor")}
 const JThread = JavaObject{Symbol("java.lang.Thread")}
 const JClassLoader = JavaObject{Symbol("java.lang.ClassLoader")}
 const JString = JavaObject{Symbol("java.lang.String")}
+const jnull = JavaObject(nothing)
+
+JavaObject(ptr::Ptr{Nothing}) = ptr == C_NULL ? JavaObject(ptr) : JavaObject{Symbol(getclassname(getclass(ptr)))}(ptr)
```
Comment (mkitti) [inline src/core.jl:141]: This depends on functions in proxy.jl at the moment, so we will probably move it into proxy.jl

**Comment by zot** [comment]
Folks, do you need any more information from me in order to tie this off?

**Comment by mkitti** [comment]
Hi @zot, thanks for checking in. JavaCall has undergone some major structural changes since a few weeks ago, so this will need to be updated again.

Next on my list is to improve JavaCall's local/global reference management. However, I think the reorganization will make things much more manageable moving forward. For example, all the `ccall` should be wrapped into the new `JNI` submodule.

Do we have your permission to continue developing this? My plan remains making this a separate package which has JavaCall as a dependency.

**Comment by zot** [comment]
> Do we have your permission to continue developing this? My plan remains making this a separate package which has JavaCall as a dependency.

Of course! JavaCall belongs to you guys -- I wrote the proxy code in order to contribute to the project. Use and modify it as you see fit, I'd just like to see it merged in one form or another...

File: src/jvm.jl (hunk 6 above)
Comment (zot) [reply src/jvm.jl:327]: Looks like init_current_vm() needs initProxy() as well

**Comment by mkitti** [comment]
@zot , @aviks I split the proxy code off into its own `JProxies` subpackage.

1. I switched most of the JNI calls over to the `JavaCall.JNI` module which wraps the `ccalls`
2. I disabled most of the memory management code since I want to use `JNI.PushLocalFrame` and `JNI.PopLocalFrame` in the future which already does the job of tracking local references and marking them for garbage collection.
3. All tests in both packages are passing.

The easiest way I found to load `JProxies` is to symlink the JProxies folder into ~/.julia/dev and then use `Pkg.develop`. More work is required, but I think this puts us on a more sustainable path.

File: JProxies/doc/index.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+## JProxy
+JProxy lets you use Java-like syntax to access fields and methods in Java. You can:
+* Get field values
+* Set field values
+* Call methods
+* Create static proxies to access static members
+Primitive types and strings are converted to Julia objects on field accesses and method returns and converted back to Java types when sent as arguments to Java methods.
+*NOTE: Because of this, if you need to call Java methods on a string that you got from Java, you'll have to use `JProxy(str)` to convert the Julia string to a proxied Java string*
+To invoke static methods, set static to true (see below).
```
Comment (zot) [inline JProxies/doc/index.md:9]: I wrote this before I added the `@class` macro -- I think it should change to this:

`To invoke static methods, use @class (see below).`

File: JProxies/doc/index.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+julia> a == b
+false
+
+julia> b == b
+true
+
+julia> JProxy(@jimport(java.lang.System)).getName()
+"java.lang.System"
+
+julia> JProxy(@jimport(java.lang.System);static=true).out.println("hello")
```
Comment (zot) [inline JProxies/doc/index.md:42]: I'd change this to
`julia> @class(java.lang.System).out.println("hello")`

**Comment by aviks** [comment]
Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?

**Comment by mkitti** [comment]
> Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?

We could do that although the direction I was heading in was to spin it out so that it can live as its own package. Currently, you could import `JProxies` without explicitly importing `JavaCall`.

```julia
using JProxies
JProxies.init()
```

Under the hood that will still use JavaCall. You made a point once that there was a lot of the proxy code to support which could make future maintenance more difficult. Dividing the code clearly like this is beneficial for maintainability I think.

There is still some code I actually want to move into JavaCall itself. For example, `listfields` probably belongs in reflect.jl.

**Comment by zot** [comment]
> > Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?
> 
> We could do that although the direction I was heading in was to spin it out so that it can live as its own package. Currently, you could import `JProxies` without explicitly importing `JavaCall`.
> 
> ```julia
> using JProxies
> JProxies.init()
> ```
> 
> Under the hood that will still use JavaCall. You made a point once that there was a lot of the proxy code to support which could make future maintenance more difficult. Dividing the code clearly like this is beneficial for maintainability I think.

That sounds like a good direction to me. That way people who just want to use JavaCall don't need to load in JProxy and people who want JProxy don't need to worry about JavaCall.

> There is still some code I actually want to move into JavaCall itself. For example, `listfields` probably belongs in reflect.jl.

Yes, I wrote some useful utility code that could move into JavaCall but I wanted to keep my changes to the base package to a minimum so I kept as much as I could in proxy.jl.

**Comment by zot** [comment]
Is this dying on the vine?

**Comment by mkitti** [comment]
Most of the changes to core JavaCall were merged in b5692cc14d7830283b90383a3b15fe752518c47e a few months ago.

We're preparing for a JavaCall 0.8 release. Taro just set compatibility with 0.8.

It looks like we have conflicts to fix here. The `jnull` thing should be tested again.

**Comment by mkitti** [comment]
Add a JProxy subpackage
